xarray/backends/api.py

```diff
@@ -264,6 +265,17 @@ def _protect_dataset_variables_inplace(dataset, cache):
             variable.data = data
 
 
+def _protect_datatree_variables_inplace(tree, cache):
+    for node in tree.subtree:
+        for name, variable in node.variables.items():
+            if name not in node._indexes:
+                # no need to protect IndexVariable objects
+                data = indexing.CopyOnWriteArray(variable._data)
+                if cache:
+                    data = indexing.MemoryCachedArray(data)
+                variable.data = data
```

Comment:
this could probably be merged with the dataset version, as you could replace `node` with `ds` here and it should still work.